File: playCribbageHand.cpp
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

#include <iostream>
#include <vector>
#include <numeric>
#include <algorithm>
#include <random>
#include <cstdlib>
#include <chrono>
#include <list>

int main(int argc, char *argv[])
{
    int nHands = 862500;
    if (argc >= 2)
    {
        nHands = atoi(argv[1]);
    }

    std::vector<int> deck(52);
    std::iota(deck.begin(), deck.end(), 0);
    // std::cout << "Deck is: ";
    // for (auto i = deck.begin(); i != deck.end(); ++i)
    //     std::cout << *i << ' ';
    // std::cout << std::endl;

    std::vector<int> deal(8);

    auto mt19937 = std::mt19937{std::random_device{}()};
    auto startTimeNs = std::chrono::high_resolution_clock::now();
    for (auto hand = 0; hand < nHands; ++hand)
    {
        std::sample(deck.begin(), deck.end(), deal.begin(), 8, mt19937);
        std::shuffle(deal.begin(), deal.end(), mt19937);
        // std::cout << "Deal is: ";
        // for (auto card = deal.begin(); card != deal.end(); ++card)
        // {
        //     std::cout << *card << ' ';
        // }
        // std::cout << std::endl;

        std::list<int> hands[] = {
            std::list<int>(deal.begin(), deal.begin() + 4),
            std::list<int>(deal.begin() + 4, deal.end())};
        // std::cout << "Hands are: ";
        // for (auto handNumber = 0; handNumber < 2; handNumber++)
        // {
        //     for (auto card = hands[handNumber].begin(); card != hands[handNumber].end(); ++card)
        //     {
        //         std::cout << *card << ' ';
        //     }
        //     std::cout << ';';
        // }
        // std::cout << std::endl;

        int playerToPlay = 0;
        while (hands[0].size() + hands[1].size() > 0)
        {
            if (hands[playerToPlay].size() > 0)
            {
                int playerToPlayPlay = hands[playerToPlay].back();
                hands[playerToPlay].pop_back();
                // std::cout << "Player" << playerToPlay << "has a play:" << playerToPlayPlay << std::endl;
            }
            playerToPlay = (playerToPlay + 1) % 2;
        }
    }

    std::chrono::duration<double> elapsedDuration = std::chrono::high_resolution_clock::now() - startTimeNs;
    auto elapsedSeconds = elapsedDuration.count();
    std::cout << "Simulated " << nHands << " hands in " << elapsedSeconds << " s for " << (elapsedSeconds * 1000000000 / nHands) << " ns per hand" << std::endl;

    return 0;
}

File: PlayCribbageHand.cs
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;

namespace MarkAFitzgerald1
{
    class PlayCribbageHand
    {
        [ThreadStatic] private static Random? random;

        static void Main(string[] args)
        {
            int totalHands = args.Length >= 1 ? Int32.Parse(args[0]) : 973000;
            int degreeOfParallelism = args.Length >= 2 ? Int32.Parse(args[1]) : Environment.ProcessorCount;
            Console.WriteLine($"Simluating {totalHands} hands with degree of parallelism {degreeOfParallelism}");
            var stopWatch = Stopwatch.StartNew();

            ParallelEnumerable.Range(0, totalHands)
                .WithDegreeOfParallelism(degreeOfParallelism)
                .ForAll((handNumber) =>
            {
                if (random == null)
                {
                    random = new Random();
                }

                var deal = DealTwoHands(random);
                // Console.WriteLine($"Deal: {string.Join(",", deal.Select(card => CardString(card)))}");
                var hands = new List<int>[] { deal.Take(4).ToList(), deal.Skip(4).ToList() };
                // Console.WriteLine($"Hands: {string.Join(",", hands[0].Select(card => CardString(card)))}; {string.Join(",", hands[1].Select(card => CardString(card)))}");
                var playerToPlay = 0;
                var playCount = 0;
                while (hands[0].Count() + hands[1].Count() > 0)
                {
                    if (hands[playerToPlay].Count() > 0)
                    {
                        var playerToPlayPlay = hands[playerToPlay][0];
                        hands[playerToPlay].RemoveAt(0);
                        playCount += CountingValue(playerToPlayPlay);
                        // Console.WriteLine($"Player {playerToPlay + 1} plays {CardString(playerToPlayPlay)} for {playCount}");
                    }
                    playerToPlay = (playerToPlay + 1) % 2;
                }
            }
            );

            stopWatch.Stop();
            // Console.WriteLine($"Stopwatch frequency = {Stopwatch.Frequency} Hz");
            Console.WriteLine($"Simulated {totalHands} hands in {1000000000L * stopWatch.ElapsedTicks / Stopwatch.Frequency} ns for {1000000000L * stopWatch.ElapsedTicks / Stopwatch.Frequency / totalHands} ns per hand");
        }

        private static List<int> DealTwoHands(Random random)
        {
            var deal = new List<int>();
            do
            {
                int item = random.Next(52);
                if (!deal.Contains(item))
                {
                    deal.Add(item);
                }
            } while (deal.Count() < 8);
            return deal;
        }

        private static int CountingValue(int playerToPlayPlay)
        {
            return Math.Min((playerToPlayPlay % 13) + 1, 10);
        }

        private static string CardString(int playerToPlayPlay)
        {
            return $"{"A23456789TJQK"[playerToPlayPlay % 13]}{"♣♦♥♠"[playerToPlayPlay / 13]}";
        }
    }
}
